feat: implement token swap functionality

parsaaba opened the pull request.

gowthamsundaresan commented on src/FinalAutoRouting.sol:
```diff
@@ -0,0 +1,2891 @@
+        if (poolAddresses.length != curveInterfaces.length)
+            revert InvalidParameter("interfaceArrays");
+
+        // Configure supported tokens
```
Can we use the same supported tokens from LTM? Since there can only be swaps between those tokens (we remain agnostic to intermediate tokens).

Because maintaining another registry here can create a new point of human error

parsaaba replied:
Got it, I see your point about eliminating dual registries. you're right that we only need to validate start/end tokens against LTM's registry.

however, there's a performance consideration: Direct mapping lookup farSupportedTokens[token] is cheaper than external calls LTM.tokenIsSupported(token) during swap validation

but for consistency, we could: 
  - Remove farSupportedTokens entirely
  - Validate tokenIn/tokenOut against LTM.tokenIsSupported()
  - Let routing logic handle intermediate tokens automatically

What do you think?

gowthamsundaresan commented on src/FinalAutoRouting.sol:
```diff
@@ -0,0 +1,2891 @@
+        require(msg.sender == address(this), "Internal only");
+
+        // Check quote freshness
+        if (block.timestamp - quote.timestamp > QUOTE_MAX_AGE)
```
Does the manager need to specifically watch out for this revert and adjust its calls accordingly (like splitting a heavy call into multiple smaller attempts, etc.)?

gowthamsundaresan commented on src/FinalAutoRouting.sol:
```diff
@@ -0,0 +1,2891 @@
+        uint256 minAmountOut
+    ) private returns (uint256 amountOut) {
+        if (params.tokenIn == address(WETH) && params.tokenOut == OSETH) {
+            return _performWETHToOsETHSwap(params, minAmountOut);
```
Is there a reason to specifically handle WETH -> osETH here instead of having the user do a multi-step swap? In case we don’t handle WETH -> osETH here, then can we remove the entire category of Protocol.MultiHop right? Since any other multi-hop is handled with (Protocol.Uniswap && route.isMultiHop)

Perhaps we can instead recommend WETH -> rETH -> osETH to the manager in its config for when it sets up RouteConfigs for AutoSwaps and put responsibility on the caller if they choose to use `swapAssets` for WETH <> osETH

gowthamsundaresan commented on src/FinalAutoRouting.sol:
```diff
@@ -0,0 +1,2891 @@
+     * @param params Swap parameters
+     * @return amountOut Amount of output tokens received by caller
+     */
+    function swapAssets(
```
I think the `params.minAmountOut` provided by the user is not actually used right?
- Quoter: checks quote -> adds buffer and calculates new minAmountOut from quote -> attempts swap -> success if outputAmount >= new minAmountOut
- Fallback: Directly calculates minAmountOut from SlippageConfig

If so, then I feel that we can give the caller the autonomy to choose their slippage without configs/fallbacks:
- Quoter: checks quote -> if quote >= params.minAmountOut -> adds buffer and calculates new minAmountOut from quote -> attempts swap -> success if outputAmount >= new minAmountOut
- Fallback: None, revert because caller has been too strict with their slippage expectations

Note 1: If this suggestion is taken, we can consider removing `SlippageConfig` altogether and instead add slippage `uint256 slippageBps` to `RouteConfig`. 

Note 2: If both suggestions from L530 and Note 1 are taken, we can consider removing `AssetType` altogether. The `bridgeAsset` can be taken in the initialization params directly OR we can add `unitOfAccount` as an immutable constant on the LiquidToken contract and use that (this actually helps in a few ways for the backend/manager too)

gowthamsundaresan commented on src/FinalAutoRouting.sol:
```diff
@@ -0,0 +1,2891 @@
+            );
+        }
+
+        if (amountOut < params.minAmountOut) revert InsufficientOutput();
```
What are your thoughts on not using `params.minAmountOut` at all and relying on the pre-existing config. So this function can v simple for the manager to call and is truly “auto”

gowthamsundaresan commented on src/FinalAutoRouting.sol:
```diff
@@ -0,0 +1,2891 @@
+        bytes32 _routePasswordHash,
+        address _liquidTokenManager,
+        bool _initializeProduction // ✅ NEW: Auto-apply production config
+    ) Ownable() {
```
Can we take `initialOwner` in the params instead of using `Ownable`? So that we follow the same pattern as the other contracts where deployer is discarded

gowthamsundaresan commented on src/FinalAutoRouting.sol:
```diff
@@ -0,0 +1,2891 @@
+            return amountOut;
+        }
+
+        // 3) Bridge route
```
Nice, this is smart!

gowthamsundaresan commented on src/FinalAutoRouting.sol:
```diff
@@ -0,0 +1,2891 @@
+     * @param password password for safety
+     * @return amountOut Amount of tokens received
+     */
+    function executeBackendSwap(
```
Can we add `minAmountOut` here and check `amountOut >= minAmountOut`?

gowthamsundaresan commented on src/FinalAutoRouting.sol:
```diff
@@ -0,0 +1,2891 @@
+            IERC20(tokenIn).safeApprove(dex, 0);
+        }
+
+        emit BackendSwapExecuted(
```
Can we include `gasUsed` like we did in `AssetSwapped`? Would be nice for backend to evaluate a given DEX over time and judge when to use it

parsaaba committed:
Add FinalAutoRoutingLib functionality

parsaaba committed:
feat: implement token swap functionality and add comprehensive test coverage

parsaaba committed:
feat: add FinalAutoRouting and interface contracts

gowthamsundaresan commented on src/FinalAutoRouting.sol:
```diff
@@ -0,0 +1,2891 @@
+    // ============================================================================
+
+    modifier onlyAuthorizedCaller() {
+        if (!authorizedCallers[msg.sender] && msg.sender != owner()) {
```
I guess the original comment is stale -- I do think it would be good to keep the option open for other contracts to call. WM may use it when settling user withdrawals (undecided as of now)

The role management style makes sense, as its in-line with the pattern of other contracts

But if we do the lib + storage method, we wouldn't need any auth on the swap functions (in lib) right? Only the management functions that control the FAR state (in storage contract) would need something like `DEFAULT_ADMIN_ROLE`?

parsaaba committed:
4th approach implementation

parsaaba committed:
Custody and ETH/Weth conversion resolved

parsaaba committed:
Superior Version of FAR with Instruction generation only

parsaaba committed:
Revised FAR

parsaaba committed:
Update FinalAutoRouting.sol

parsaaba committed:
Path Generator version of FAR added

parsaaba committed:
optimized and inner audit version added

parsaaba committed:
Resolve audit issues

parsaaba committed:
optimized and inner audit version added2

parsaaba committed:
initial version of LTM

parsaaba committed:
feat: implement token swap functionality and refactor tests

parsaaba committed:
refactor: update FinalAutoRouting and LiquidTokenManager contracts

- Improve ETH transfer safety in emergencyWithdraw with proper error handling
- Clean up comments and imports in FinalAutoRouting
- Update LiquidTokenManager with latest implementation
- Remove deprecated TokenSwap test directory and related test files

parsaaba committed:
refactor: improve code formatting and Update LTM script files to use payable address type for LiquidTokenManager

gowthamsundaresan commented on src/core/LiquidTokenManager.sol:
```diff
@@ -337,6 +360,194 @@ contract LiquidTokenManager is
+
+        // Swap using FAR - for every tokenIn swap to corresponding tokenOut
+        for (uint256 i = 0; i < assetsLength; i++) {
+            address tokenIn = address(assetsToSwap[i]);
```
Note that `tokenIn` must be a supported token and have a corresponding supported strategy

gowthamsundaresan commented:
Hey putting this PR into draft as we are splitting it up into LTM integration and a new repo. Will keep it open for reference and close it entirely after merge of `v2-core` to `dev`

gowthamsundaresan commented:
Hey putting this PR into draft as we are splitting it up into LTM integration and a new repo. Will keep it open for reference and close it entirely after merge of v2-core to dev
